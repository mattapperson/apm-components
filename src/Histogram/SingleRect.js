import React from 'react';
import PropTypes from 'prop-types';

function SingleRect({ innerHeight, marginTop, style, x, width }) {
  return (
    <g transform={`translate(0, ${marginTop})`}>
      <rect
        style={style}
        height={innerHeight}
        width={width}
        rx={'2px'}
        ry={'2px'}
        x={x}
      />
    </g>
  );
}

SingleRect.requiresSVG = true;
SingleRect.propTypes = {
  x: PropTypes.number.isRequired
};

export default SingleRect;
