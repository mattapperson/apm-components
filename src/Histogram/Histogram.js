import React, { PureComponent } from 'react';
import { scaleLinear } from 'd3-scale';
import SingleRect from './SingleRect';
import 'react-vis/dist/style.css';
import {
  XYPlot,
  XAxis,
  YAxis,
  HorizontalGridLines,
  VerticalRectSeries,
  Voronoi
} from 'react-vis';
import { getYMax, getXMax, getYMaxRounded } from '../chart_utils';

const XY_HEIGHT = 120;
const XY_WIDTH = 900;
const XY_MARGIN = {
  top: 20,
  left: 50,
  right: 10
};

const NUM_OF_X_TICK = 10;

class Histogram extends PureComponent {
  constructor(props) {
    super(props);
    this.state = {
      hoveredBucket: null
    };
  }

  onClick = bucket => {
    return this.props.buckets[bucket.i].y > 0 && this.props.onClick(bucket.i);
  };

  onHover = bucket => {
    this.props.buckets[bucket.i].y > 0 &&
      this.setState({ hoveredBucket: bucket });
  };

  onBlur = () => {
    this.setState({ hoveredBucket: null });
  };

  getWithHighlightedBucket(items, selected) {
    return items
      .map((item, i) => {
        if (i === selected) {
          return { ...item, color: '#3360a3' };
        }
        return item;
      })
      .map((item, i) => {
        const padding = (item.x - item.x0) / 20;
        return { ...item, x: item.x - padding, x0: item.x0 + padding, i };
      });
  }

  render() {
    const { buckets, selectedBucket, bucketSize } = this.props;

    if (!buckets) {
      return null;
    }

    const xMin = 0;
    const xMax = getXMax(buckets);
    const yMin = 0;
    const yMax = getYMax(buckets);
    const yMaxRounded = getYMaxRounded(yMax);
    const yTickValues = [yMaxRounded, yMaxRounded / 2];

    const x = scaleLinear()
      .domain([xMin, xMax])
      .range([XY_MARGIN.left, XY_WIDTH - XY_MARGIN.right]);
    const y = scaleLinear().domain([yMin, yMaxRounded]).range([XY_HEIGHT, 0]);

    return (
      <XYPlot
        width={XY_WIDTH}
        height={XY_HEIGHT}
        margin={XY_MARGIN}
        xDomain={x.domain()}
        yDomain={y.domain()}
      >
        <HorizontalGridLines tickValues={yTickValues} />
        <XAxis
          style={{ strokeWidth: '1px' }}
          marginRight={10}
          tickSizeOuter={10}
          tickSizeInner={0}
          tickTotal={NUM_OF_X_TICK}
        />
        <YAxis
          tickSize={0}
          hideLine
          marginTop={XY_MARGIN.top}
          tickValues={yTickValues}
          tickFormat={y => {
            return `${y} reqs.`;
          }}
        />

        {this.state.hoveredBucket
          ? <SingleRect
              x={x(this.state.hoveredBucket.x0)}
              width={x(bucketSize) - x(0)}
              marginTop={XY_MARGIN.top}
              style={{
                fill: '#dddddd'
              }}
            />
          : null}

        {Number.isInteger(selectedBucket)
          ? <SingleRect
              x={x(selectedBucket * bucketSize)}
              width={x(bucketSize) - x(0)}
              marginTop={XY_MARGIN.top}
              style={{
                fill: 'transparent',
                stroke: 'rgb(172, 189, 220)'
              }}
            />
          : null}

        <VerticalRectSeries
          colorType="literal"
          color="rgb(172, 189, 216)"
          data={this.getWithHighlightedBucket(buckets, selectedBucket)}
          style={{
            rx: '2px',
            ry: '2px'
          }}
        />

        <Voronoi
          extent={[[XY_MARGIN.left, XY_MARGIN.top], [XY_WIDTH, XY_HEIGHT]]}
          nodes={this.props.buckets.map(item => ({
            ...item,
            x: (item.x0 + item.x) / 2,
            y: 1
          }))}
          onClick={this.onClick}
          onHover={this.onHover}
          onBlur={this.onBlur}
          x={d => x(d.x)}
          y={d => y(d.y)}
        />
      </XYPlot>
    );
  }
}

export default Histogram;
