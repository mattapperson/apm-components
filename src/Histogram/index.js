import React from 'react';
import Histogram from './Histogram';
import response from './response.json';

const buckets = getFormattedBuckets(response.buckets, response.bucketSize);

function getFormattedBuckets(buckets, bucketSize) {
  if (!buckets) {
    return null;
  }

  const yMax = Math.max(...buckets.map(item => item.count));
  const yMin = yMax * 0.1;

  return buckets.map(({ count, key, transactionId }) => {
    return {
      transactionId,
      x0: key,
      x: key + bucketSize,
      y: count > 0 ? Math.max(count, yMin) : 0
    };
  });
}

export default class extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      transactionId: null
    };
  }

  render() {
    return (
      <Histogram
        buckets={buckets}
        bucketSize={response.bucketSize}
        transactionId={this.state.transactionId}
        onClick={selectedBucket => {
          this.setState({ transactionId: selectedBucket.transactionId });
        }}
      />
    );
  }
}
