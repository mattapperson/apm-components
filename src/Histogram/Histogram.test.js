import React from 'react';
import { mount } from 'enzyme';
import toDiffableHtml from 'diffable-html';
import { HistogramWithoutHOC } from './Histogram';
import { getFormattedBuckets } from './index';
import response from './response.json';

describe('Histogram', () => {
  let wrapper;
  const onClick = jest.fn();

  beforeEach(() => {
    const buckets = getFormattedBuckets(response.buckets, response.bucketSize);
    wrapper = mount(
      <HistogramWithoutHOC
        buckets={buckets}
        bucketSize={response.bucketSize}
        transactionId="myTransactionId"
        onClick={onClick}
        formatXValue={value => `${value} test`}
        formatYValue={value => `${value} rpm`}
        formatTooltipHeader={(hoveredX0, hoveredX) =>
          `${hoveredX0 / 1000} - ${hoveredX / 1000} ms`}
        tooltipLegendTitle="Requests"
        width={800}
      />
    );
  });

  describe('Initially', () => {
    it('should have default state', () => {
      expect(wrapper.state()).toEqual({ hoveredBucket: null });
    });

    it('should have default markup', () => {
      expect(toDiffableHtml(wrapper.html())).toMatchSnapshot();
    });

    it('should not show tooltip', () => {
      expect(wrapper.find('Tooltip').length).toBe(0);
    });
  });

  describe('when hovering over an empty bucket', () => {
    beforeEach(() => {
      wrapper
        .find('.rv-voronoi__cell')
        .at(2)
        .simulate('mouseOver');
    });

    it('should not display tooltip', () => {
      expect(wrapper.find('Tooltip').length).toBe(0);
    });
  });

  describe('when hovering over a non-empty bucket', () => {
    beforeEach(() => {
      wrapper
        .find('.rv-voronoi__cell')
        .at(7)
        .simulate('mouseOver');
    });

    it('should display tooltip', () => {
      const tooltips = wrapper.find('Tooltip');

      expect(tooltips.length).toBe(1);
      expect(tooltips.prop('header')).toBe('811.076 - 869.01 ms');
      expect(tooltips.prop('tooltipPoints')).toEqual([
        { color: 'rgb(172, 189, 216)', text: 'Requests', value: 49 }
      ]);
      expect(tooltips.prop('x')).toEqual(869010);
      expect(tooltips.prop('y')).toEqual(27.5);
    });

    it('should update state with "hoveredBucket"', () => {
      expect(wrapper.state()).toEqual({
        hoveredBucket: {
          transactionId: '99c50a5b-44b4-4289-a3d1-a2815d128192',
          x: 869010,
          x0: 811076,
          y: 49
        }
      });
    });

    it('should have correct markup for tooltip', () => {
      const tooltips = wrapper.find('Tooltip');
      expect(toDiffableHtml(tooltips.html())).toMatchSnapshot();
    });
  });

  describe('when clicking on a non-empty bucket', () => {
    beforeEach(() => {
      wrapper
        .find('.rv-voronoi__cell')
        .at(7)
        .simulate('click');
    });

    it('should call onClick with bucket', () => {
      expect(onClick).toHaveBeenCalledWith({
        transactionId: '99c50a5b-44b4-4289-a3d1-a2815d128192',
        x: 869010,
        x0: 811076,
        y: 49
      });
    });
  });
});
