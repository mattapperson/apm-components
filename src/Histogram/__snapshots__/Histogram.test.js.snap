// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Histogram Initially should have default markup 1`] = `
"
<div class=\\"sc-htoDjs gBxNsV\\">
  <div class=\\"rv-xy-plot \\"
       style=\\"width: 800px; height: 128px;\\"
  >
    <svg class=\\"rv-xy-plot__inner\\"
         width=\\"800\\"
         height=\\"128\\"
    >
      <g transform=\\"translate(80,16)\\"
         class=\\"rv-xy-plot__grid-lines\\"
      >
        <line y1=\\"80\\"
              y2=\\"80\\"
              x1=\\"0\\"
              x2=\\"704\\"
              class=\\"rv-xy-plot__grid-lines__line\\"
        >
        </line>
        <line y1=\\"40\\"
              y2=\\"40\\"
              x1=\\"0\\"
              x2=\\"704\\"
              class=\\"rv-xy-plot__grid-lines__line\\"
        >
        </line>
        <line y1=\\"0\\"
              y2=\\"0\\"
              x1=\\"0\\"
              x2=\\"704\\"
              class=\\"rv-xy-plot__grid-lines__line\\"
        >
        </line>
      </g>
      <g transform=\\"translate(80,96)\\"
         class=\\"rv-xy-plot__axis rv-xy-plot__axis--horizontal \\"
         style=\\"stroke-width: 1px;\\"
      >
        <line x1=\\"0\\"
              x2=\\"704\\"
              y1=\\"0\\"
              y2=\\"0\\"
              class=\\"rv-xy-plot__axis__line\\"
              style=\\"stroke-width: 1px;\\"
        >
        </line>
        <g transform=\\"translate(0, 0)\\"
           class=\\"rv-xy-plot__axis__ticks\\"
        >
          <g transform=\\"translate(0, 0)\\"
             class=\\"rv-xy-plot__axis__tick\\"
             style=\\"stroke-width: 1px;\\"
          >
            <line x1=\\"0\\"
                  x2=\\"0\\"
                  y1=\\"0\\"
                  y2=\\"10\\"
                  class=\\"rv-xy-plot__axis__tick__line\\"
                  style=\\"stroke-width: 1px;\\"
            >
            </line>
            <text text-anchor=\\"middle\\"
                  dy=\\"0.72em\\"
                  transform=\\"translate(0, 18)\\"
                  class=\\"rv-xy-plot__axis__tick__text\\"
                  style=\\"stroke-width: 1px;\\"
            >
              0 ms
            </text>
          </g>
          <g transform=\\"translate(108.49784730407508, 0)\\"
             class=\\"rv-xy-plot__axis__tick\\"
             style=\\"stroke-width: 1px;\\"
          >
            <line x1=\\"0\\"
                  x2=\\"0\\"
                  y1=\\"0\\"
                  y2=\\"10\\"
                  class=\\"rv-xy-plot__axis__tick__line\\"
                  style=\\"stroke-width: 1px;\\"
            >
            </line>
            <text text-anchor=\\"middle\\"
                  dy=\\"0.72em\\"
                  transform=\\"translate(0, 18)\\"
                  class=\\"rv-xy-plot__axis__tick__text\\"
                  style=\\"stroke-width: 1px;\\"
            >
              500 ms
            </text>
          </g>
          <g transform=\\"translate(216.99569460815016, 0)\\"
             class=\\"rv-xy-plot__axis__tick\\"
             style=\\"stroke-width: 1px;\\"
          >
            <line x1=\\"0\\"
                  x2=\\"0\\"
                  y1=\\"0\\"
                  y2=\\"10\\"
                  class=\\"rv-xy-plot__axis__tick__line\\"
                  style=\\"stroke-width: 1px;\\"
            >
            </line>
            <text text-anchor=\\"middle\\"
                  dy=\\"0.72em\\"
                  transform=\\"translate(0, 18)\\"
                  class=\\"rv-xy-plot__axis__tick__text\\"
                  style=\\"stroke-width: 1px;\\"
            >
              1,000 ms
            </text>
          </g>
          <g transform=\\"translate(325.4935419122253, 0)\\"
             class=\\"rv-xy-plot__axis__tick\\"
             style=\\"stroke-width: 1px;\\"
          >
            <line x1=\\"0\\"
                  x2=\\"0\\"
                  y1=\\"0\\"
                  y2=\\"10\\"
                  class=\\"rv-xy-plot__axis__tick__line\\"
                  style=\\"stroke-width: 1px;\\"
            >
            </line>
            <text text-anchor=\\"middle\\"
                  dy=\\"0.72em\\"
                  transform=\\"translate(0, 18)\\"
                  class=\\"rv-xy-plot__axis__tick__text\\"
                  style=\\"stroke-width: 1px;\\"
            >
              1,500 ms
            </text>
          </g>
          <g transform=\\"translate(433.9913892163003, 0)\\"
             class=\\"rv-xy-plot__axis__tick\\"
             style=\\"stroke-width: 1px;\\"
          >
            <line x1=\\"0\\"
                  x2=\\"0\\"
                  y1=\\"0\\"
                  y2=\\"10\\"
                  class=\\"rv-xy-plot__axis__tick__line\\"
                  style=\\"stroke-width: 1px;\\"
            >
            </line>
            <text text-anchor=\\"middle\\"
                  dy=\\"0.72em\\"
                  transform=\\"translate(0, 18)\\"
                  class=\\"rv-xy-plot__axis__tick__text\\"
                  style=\\"stroke-width: 1px;\\"
            >
              2,000 ms
            </text>
          </g>
          <g transform=\\"translate(542.4892365203754, 0)\\"
             class=\\"rv-xy-plot__axis__tick\\"
             style=\\"stroke-width: 1px;\\"
          >
            <line x1=\\"0\\"
                  x2=\\"0\\"
                  y1=\\"0\\"
                  y2=\\"10\\"
                  class=\\"rv-xy-plot__axis__tick__line\\"
                  style=\\"stroke-width: 1px;\\"
            >
            </line>
            <text text-anchor=\\"middle\\"
                  dy=\\"0.72em\\"
                  transform=\\"translate(0, 18)\\"
                  class=\\"rv-xy-plot__axis__tick__text\\"
                  style=\\"stroke-width: 1px;\\"
            >
              2,500 ms
            </text>
          </g>
          <g transform=\\"translate(650.9870838244506, 0)\\"
             class=\\"rv-xy-plot__axis__tick\\"
             style=\\"stroke-width: 1px;\\"
          >
            <line x1=\\"0\\"
                  x2=\\"0\\"
                  y1=\\"0\\"
                  y2=\\"10\\"
                  class=\\"rv-xy-plot__axis__tick__line\\"
                  style=\\"stroke-width: 1px;\\"
            >
            </line>
            <text text-anchor=\\"middle\\"
                  dy=\\"0.72em\\"
                  transform=\\"translate(0, 18)\\"
                  class=\\"rv-xy-plot__axis__tick__text\\"
                  style=\\"stroke-width: 1px;\\"
            >
              3,000 ms
            </text>
          </g>
        </g>
      </g>
      <g transform=\\"translate(0,16)\\"
         class=\\"rv-xy-plot__axis rv-xy-plot__axis--vertical \\"
      >
        <g transform=\\"translate(80, 0)\\"
           class=\\"rv-xy-plot__axis__ticks\\"
        >
          <g transform=\\"translate(0, 80)\\"
             class=\\"rv-xy-plot__axis__tick\\"
          >
            <line y1=\\"0\\"
                  y2=\\"0\\"
                  x1=\\"0\\"
                  x2=\\"0\\"
                  class=\\"rv-xy-plot__axis__tick__line\\"
            >
            </line>
            <text text-anchor=\\"end\\"
                  dy=\\"0.32em\\"
                  transform=\\"translate(-8, 0)\\"
                  class=\\"rv-xy-plot__axis__tick__text\\"
            >
              0.0
            </text>
          </g>
          <g transform=\\"translate(0, 40)\\"
             class=\\"rv-xy-plot__axis__tick\\"
          >
            <line y1=\\"0\\"
                  y2=\\"0\\"
                  x1=\\"0\\"
                  x2=\\"0\\"
                  class=\\"rv-xy-plot__axis__tick__line\\"
            >
            </line>
            <text text-anchor=\\"end\\"
                  dy=\\"0.32em\\"
                  transform=\\"translate(-8, 0)\\"
                  class=\\"rv-xy-plot__axis__tick__text\\"
            >
              27.5
            </text>
          </g>
          <g transform=\\"translate(0, 0)\\"
             class=\\"rv-xy-plot__axis__tick\\"
          >
            <line y1=\\"0\\"
                  y2=\\"0\\"
                  x1=\\"0\\"
                  x2=\\"0\\"
                  class=\\"rv-xy-plot__axis__tick__line\\"
            >
            </line>
            <text text-anchor=\\"end\\"
                  dy=\\"0.32em\\"
                  transform=\\"translate(-8, 0)\\"
                  class=\\"rv-xy-plot__axis__tick__text\\"
            >
              55.0
            </text>
          </g>
        </g>
      </g>
      <g class=\\"rv-xy-plot__series rv-xy-plot__series--rect \\"
         transform=\\"translate(80,16)\\"
      >
        <rect x=\\"1.2571428571428571\\"
              width=\\"22.628571428571433\\"
              y=\\"80\\"
              height=\\"0\\"
              style=\\"opacity: 1; stroke: rgb(172, 189, 216); fill: rgb(172, 189, 216);\\"
        >
        </rect>
        <rect x=\\"26.4\\"
              width=\\"22.628571428571433\\"
              y=\\"80\\"
              height=\\"0\\"
              style=\\"opacity: 1; stroke: rgb(172, 189, 216); fill: rgb(172, 189, 216);\\"
        >
        </rect>
        <rect x=\\"51.542857142857144\\"
              width=\\"22.62857142857142\\"
              y=\\"80\\"
              height=\\"0\\"
              style=\\"opacity: 1; stroke: rgb(172, 189, 216); fill: rgb(172, 189, 216);\\"
        >
        </rect>
        <rect x=\\"76.68571428571428\\"
              width=\\"22.628571428571433\\"
              y=\\"80\\"
              height=\\"0\\"
              style=\\"opacity: 1; stroke: rgb(172, 189, 216); fill: rgb(172, 189, 216);\\"
        >
        </rect>
        <rect x=\\"101.82857142857144\\"
              width=\\"22.628571428571405\\"
              y=\\"80\\"
              height=\\"0\\"
              style=\\"opacity: 1; stroke: rgb(172, 189, 216); fill: rgb(172, 189, 216);\\"
        >
        </rect>
        <rect x=\\"126.97142857142858\\"
              width=\\"22.62857142857142\\"
              y=\\"68.36363636363636\\"
              height=\\"11.63636363636364\\"
              style=\\"opacity: 1; stroke: rgb(172, 189, 216); fill: rgb(172, 189, 216);\\"
        >
        </rect>
        <rect x=\\"152.11428571428573\\"
              width=\\"22.628571428571405\\"
              y=\\"46.54545454545455\\"
              height=\\"33.45454545454545\\"
              style=\\"opacity: 1; stroke: rgb(172, 189, 216); fill: rgb(172, 189, 216);\\"
        >
        </rect>
        <rect x=\\"177.25714285714287\\"
              width=\\"22.628571428571405\\"
              y=\\"8.727272727272734\\"
              height=\\"71.27272727272727\\"
              style=\\"opacity: 1; stroke: rgb(172, 189, 216); fill: rgb(172, 189, 216);\\"
        >
        </rect>
        <rect x=\\"202.40000000000003\\"
              width=\\"22.628571428571377\\"
              y=\\"5.818181818181813\\"
              height=\\"74.18181818181819\\"
              style=\\"opacity: 1; stroke: rgb(172, 189, 216); fill: rgb(172, 189, 216);\\"
        >
        </rect>
        <rect x=\\"227.54285714285712\\"
              width=\\"22.628571428571462\\"
              y=\\"13.090909090909093\\"
              height=\\"66.9090909090909\\"
              style=\\"opacity: 1; stroke: rgb(172, 189, 216); fill: rgb(172, 189, 216);\\"
        >
        </rect>
        <rect x=\\"252.68571428571425\\"
              width=\\"22.62857142857149\\"
              y=\\"61.09090909090909\\"
              height=\\"18.909090909090907\\"
              style=\\"opacity: 1; stroke: rgb(172, 189, 216); fill: rgb(172, 189, 216);\\"
        >
        </rect>
        <rect x=\\"277.8285714285714\\"
              width=\\"22.62857142857149\\"
              y=\\"69.81818181818181\\"
              height=\\"10.181818181818187\\"
              style=\\"opacity: 1; stroke: rgb(172, 189, 216); fill: rgb(172, 189, 216);\\"
        >
        </rect>
        <rect x=\\"302.97142857142853\\"
              width=\\"22.62857142857149\\"
              y=\\"72.58181818181818\\"
              height=\\"7.4181818181818215\\"
              style=\\"opacity: 1; stroke: rgb(172, 189, 216); fill: rgb(172, 189, 216);\\"
        >
        </rect>
        <rect x=\\"328.1142857142857\\"
              width=\\"22.628571428571433\\"
              y=\\"72.58181818181818\\"
              height=\\"7.4181818181818215\\"
              style=\\"opacity: 1; stroke: rgb(172, 189, 216); fill: rgb(172, 189, 216);\\"
        >
        </rect>
        <rect x=\\"353.25714285714287\\"
              width=\\"22.628571428571433\\"
              y=\\"80\\"
              height=\\"0\\"
              style=\\"opacity: 1; stroke: rgb(172, 189, 216); fill: rgb(172, 189, 216);\\"
        >
        </rect>
        <rect x=\\"378.4\\"
              width=\\"22.628571428571433\\"
              y=\\"72.58181818181818\\"
              height=\\"7.4181818181818215\\"
              style=\\"opacity: 1; stroke: rgb(172, 189, 216); fill: rgb(172, 189, 216);\\"
        >
        </rect>
        <rect x=\\"403.5428571428571\\"
              width=\\"22.628571428571547\\"
              y=\\"80\\"
              height=\\"0\\"
              style=\\"opacity: 1; stroke: rgb(172, 189, 216); fill: rgb(172, 189, 216);\\"
        >
        </rect>
        <rect x=\\"428.68571428571425\\"
              width=\\"22.62857142857149\\"
              y=\\"80\\"
              height=\\"0\\"
              style=\\"opacity: 1; stroke: rgb(172, 189, 216); fill: rgb(172, 189, 216);\\"
        >
        </rect>
        <rect x=\\"453.82857142857137\\"
              width=\\"22.628571428571547\\"
              y=\\"72.58181818181818\\"
              height=\\"7.4181818181818215\\"
              style=\\"opacity: 1; stroke: rgb(172, 189, 216); fill: rgb(172, 189, 216);\\"
        >
        </rect>
        <rect x=\\"478.9714285714286\\"
              width=\\"22.628571428571433\\"
              y=\\"80\\"
              height=\\"0\\"
              style=\\"opacity: 1; stroke: rgb(172, 189, 216); fill: rgb(172, 189, 216);\\"
        >
        </rect>
        <rect x=\\"504.1142857142857\\"
              width=\\"22.628571428571433\\"
              y=\\"80\\"
              height=\\"0\\"
              style=\\"opacity: 1; stroke: rgb(172, 189, 216); fill: rgb(172, 189, 216);\\"
        >
        </rect>
        <rect x=\\"529.2571428571429\\"
              width=\\"22.628571428571377\\"
              y=\\"80\\"
              height=\\"0\\"
              style=\\"opacity: 1; stroke: rgb(172, 189, 216); fill: rgb(172, 189, 216);\\"
        >
        </rect>
        <rect x=\\"554.4\\"
              width=\\"22.62857142857149\\"
              y=\\"80\\"
              height=\\"0\\"
              style=\\"opacity: 1; stroke: rgb(172, 189, 216); fill: rgb(172, 189, 216);\\"
        >
        </rect>
        <rect x=\\"579.5428571428571\\"
              width=\\"22.62857142857149\\"
              y=\\"80\\"
              height=\\"0\\"
              style=\\"opacity: 1; stroke: rgb(172, 189, 216); fill: rgb(172, 189, 216);\\"
        >
        </rect>
        <rect x=\\"604.6857142857142\\"
              width=\\"22.628571428571604\\"
              y=\\"72.58181818181818\\"
              height=\\"7.4181818181818215\\"
              style=\\"opacity: 1; stroke: rgb(172, 189, 216); fill: rgb(172, 189, 216);\\"
        >
        </rect>
        <rect x=\\"629.8285714285714\\"
              width=\\"22.62857142857149\\"
              y=\\"72.58181818181818\\"
              height=\\"7.4181818181818215\\"
              style=\\"opacity: 1; stroke: rgb(172, 189, 216); fill: rgb(172, 189, 216);\\"
        >
        </rect>
        <rect x=\\"654.9714285714285\\"
              width=\\"22.62857142857149\\"
              y=\\"80\\"
              height=\\"0\\"
              style=\\"opacity: 1; stroke: rgb(172, 189, 216); fill: rgb(172, 189, 216);\\"
        >
        </rect>
        <rect x=\\"680.1142857142858\\"
              width=\\"22.628571428571377\\"
              y=\\"72.58181818181818\\"
              height=\\"7.4181818181818215\\"
              style=\\"opacity: 1; stroke: rgb(172, 189, 216); fill: rgb(172, 189, 216);\\"
        >
        </rect>
      </g>
      <g class=\\" rv-voronoi\\">
        <path class=\\"rv-voronoi__cell\\"
              d=\\"M105.14285749999999,128L105.14285749999999,16L80,16L80,128Z\\"
              fill=\\"none\\"
              style=\\"pointer-events: all;\\"
        >
        </path>
        <path class=\\"rv-voronoi__cell\\"
              d=\\"M105.14285749999999,16L105.14285749999999,128L130.28571449999998,128L130.28571449999998,16Z\\"
              fill=\\"none\\"
              style=\\"pointer-events: all;\\"
        >
        </path>
        <path class=\\"rv-voronoi__cell\\"
              d=\\"M130.28571449999998,16L130.28571449999998,128L155.42857149999998,128L155.42857149999998,16Z\\"
              fill=\\"none\\"
              style=\\"pointer-events: all;\\"
        >
        </path>
        <path class=\\"rv-voronoi__cell\\"
              d=\\"M155.42857149999998,16L155.42857149999998,128L180.5714285,128L180.5714285,16Z\\"
              fill=\\"none\\"
              style=\\"pointer-events: all;\\"
        >
        </path>
        <path class=\\"rv-voronoi__cell\\"
              d=\\"M180.5714285,16L180.5714285,128L205.7142855,128L205.7142855,16Z\\"
              fill=\\"none\\"
              style=\\"pointer-events: all;\\"
        >
        </path>
        <path class=\\"rv-voronoi__cell\\"
              d=\\"M205.7142855,16L205.7142855,128L230.85714249999998,128L230.85714249999998,16Z\\"
              fill=\\"none\\"
              style=\\"pointer-events: all;\\"
        >
        </path>
        <path class=\\"rv-voronoi__cell\\"
              d=\\"M230.85714249999998,16L230.85714249999998,128L255.99999999999997,128L255.99999999999997,16Z\\"
              fill=\\"none\\"
              style=\\"pointer-events: all;\\"
        >
        </path>
        <path class=\\"rv-voronoi__cell\\"
              d=\\"M255.99999999999997,16L255.99999999999997,128L281.1428575,128L281.1428575,16Z\\"
              fill=\\"none\\"
              style=\\"pointer-events: all;\\"
        >
        </path>
        <path class=\\"rv-voronoi__cell\\"
              d=\\"M281.1428575,16L281.1428575,128L306.2857145,128L306.2857145,16Z\\"
              fill=\\"none\\"
              style=\\"pointer-events: all;\\"
        >
        </path>
        <path class=\\"rv-voronoi__cell\\"
              d=\\"M306.2857145,16L306.2857145,128L331.4285715,128L331.4285715,16Z\\"
              fill=\\"none\\"
              style=\\"pointer-events: all;\\"
        >
        </path>
        <path class=\\"rv-voronoi__cell\\"
              d=\\"M331.4285715,16L331.4285715,128L356.5714285,128L356.5714285,16Z\\"
              fill=\\"none\\"
              style=\\"pointer-events: all;\\"
        >
        </path>
        <path class=\\"rv-voronoi__cell\\"
              d=\\"M356.5714285,16L356.5714285,128L381.71428549999996,128L381.71428549999996,16Z\\"
              fill=\\"none\\"
              style=\\"pointer-events: all;\\"
        >
        </path>
        <path class=\\"rv-voronoi__cell\\"
              d=\\"M381.71428549999996,16L381.71428549999996,128L406.8571425,128L406.8571425,16Z\\"
              fill=\\"none\\"
              style=\\"pointer-events: all;\\"
        >
        </path>
        <path class=\\"rv-voronoi__cell\\"
              d=\\"M406.8571425,16L406.8571425,128L432,128L432,16Z\\"
              fill=\\"none\\"
              style=\\"pointer-events: all;\\"
        >
        </path>
        <path class=\\"rv-voronoi__cell\\"
              d=\\"M432,16L432,128L457.1428575,128L457.1428575,16Z\\"
              fill=\\"none\\"
              style=\\"pointer-events: all;\\"
        >
        </path>
        <path class=\\"rv-voronoi__cell\\"
              d=\\"M457.1428575,16L457.1428575,128L482.2857144999999,128L482.2857144999999,16Z\\"
              fill=\\"none\\"
              style=\\"pointer-events: all;\\"
        >
        </path>
        <path class=\\"rv-voronoi__cell\\"
              d=\\"M482.2857144999999,16L482.2857144999999,128L507.4285715,128L507.4285715,16Z\\"
              fill=\\"none\\"
              style=\\"pointer-events: all;\\"
        >
        </path>
        <path class=\\"rv-voronoi__cell\\"
              d=\\"M507.4285715,16L507.4285715,128L532.5714284999999,128L532.5714284999999,16Z\\"
              fill=\\"none\\"
              style=\\"pointer-events: all;\\"
        >
        </path>
        <path class=\\"rv-voronoi__cell\\"
              d=\\"M532.5714284999999,16L532.5714284999999,128L557.7142855,128L557.7142855,16Z\\"
              fill=\\"none\\"
              style=\\"pointer-events: all;\\"
        >
        </path>
        <path class=\\"rv-voronoi__cell\\"
              d=\\"M557.7142855,16L557.7142855,128L582.8571425,128L582.8571425,16Z\\"
              fill=\\"none\\"
              style=\\"pointer-events: all;\\"
        >
        </path>
        <path class=\\"rv-voronoi__cell\\"
              d=\\"M582.8571425,16L582.8571425,128L608,128L608,16Z\\"
              fill=\\"none\\"
              style=\\"pointer-events: all;\\"
        >
        </path>
        <path class=\\"rv-voronoi__cell\\"
              d=\\"M608,16L608,128L633.1428575,128L633.1428575,16Z\\"
              fill=\\"none\\"
              style=\\"pointer-events: all;\\"
        >
        </path>
        <path class=\\"rv-voronoi__cell\\"
              d=\\"M633.1428575,16L633.1428575,128L658.2857145,128L658.2857145,16Z\\"
              fill=\\"none\\"
              style=\\"pointer-events: all;\\"
        >
        </path>
        <path class=\\"rv-voronoi__cell\\"
              d=\\"M658.2857145,16L658.2857145,128L683.4285715,128L683.4285715,16Z\\"
              fill=\\"none\\"
              style=\\"pointer-events: all;\\"
        >
        </path>
        <path class=\\"rv-voronoi__cell\\"
              d=\\"M683.4285715,16L683.4285715,128L708.5714284999999,128L708.5714284999999,16Z\\"
              fill=\\"none\\"
              style=\\"pointer-events: all;\\"
        >
        </path>
        <path class=\\"rv-voronoi__cell\\"
              d=\\"M708.5714284999999,16L708.5714284999999,128L733.7142855,128L733.7142855,16Z\\"
              fill=\\"none\\"
              style=\\"pointer-events: all;\\"
        >
        </path>
        <path class=\\"rv-voronoi__cell\\"
              d=\\"M733.7142855,16L733.7142855,128L758.8571425,128L758.8571425,16Z\\"
              fill=\\"none\\"
              style=\\"pointer-events: all;\\"
        >
        </path>
        <path class=\\"rv-voronoi__cell\\"
              d=\\"M758.8571425,16L758.8571425,128L800,128L800,16Z\\"
              fill=\\"none\\"
              style=\\"pointer-events: all;\\"
        >
        </path>
      </g>
    </svg>
  </div>
</div>
"
`;

exports[`Histogram when hovering over a non-empty bucket should have correct markup for tooltip 1`] = `
"
<div class=\\"rv-hint  rv-hint--horizontalAlign-right
     rv-hint--verticalAlign-bottom\\"
     style=\\"margin-left: 1%; margin-right: 1%; left: 268.57142857142856px; top: 56px; position: absolute;\\"
>
  <div class=\\"sc-htpNat jejoiB\\">
    <div class=\\"sc-bxivhb etAnDE\\">
      811 - 869 ms
    </div>
    <div class=\\"sc-ifAKCX gVdyMJ\\">
      <div class=\\"sc-EHOje hxHpbH\\">
        <div class=\\"sc-bZQynM bHRdln sc-bdVaJa eWvHJG\\"
             font-size=\\"12px\\"
        >
          <span class=\\"sc-bwzfXH eKfWHj\\"
                color=\\"rgb(172, 189, 216)\\"
                radius=\\"8\\"
          >
          </span>
          <!-- react-text: 109 -->
          Requests
          <!-- /react-text -->
        </div>
        <div class=\\"sc-gzVnrw zwJEr\\">
          49.0
        </div>
      </div>
    </div>
  </div>
</div>
"
`;
